newFOROP: fix crash when optimizing 2-var for over builtin::indexed

mauke opened the pull request:
OP_ENTERSUB isn't necessarily a LISTOP, apparently, so we can't just grab its op_last. Instead, copy/paste logic from elsewhere in op.c to find the cvop.

Also, avoid crashing on "fake" pad entries that represent lexical subs from outer scopes by climbing up the scope chain until we reach a real pad entry.

Fixes #23405.


---------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and it is included.

mauke committed:
newFOROP: fix crash when optimizing 2-var for over builtin::indexed

OP_ENTERSUB isn't necessarily a LISTOP, apparently, so we can't just
grab its op_last. Instead, copy/paste logic from elsewhere in op.c to
find the cvop.

Also, avoid crashing on "fake" pad entries that represent lexical subs
from outer scopes by climbing up the scope chain until we reach a real
pad entry.

Fixes #23405.

mauke committed:
perldelta for the 2-var 'for' crash (#23405)

tonycoz commented on t/perf/opcount.t:
```diff
@@ -698,6 +698,21 @@ test_opcount(0, "multiconcat: local assign",
 
 # builtin:: function calls should be replaced with efficient op implementations
 no warnings 'experimental::builtin';
+use builtin qw(
+    blessed
+    ceil
+    false
+    floor
```
Most of the changes in this file seem like they belong in a separate commit - they have nothing to do with fixing the builtin::indexed optimization

mauke replied:
The builtin::indexed optimization works fine on `builtin::indexed(...)`, but may crash on `use builtin qw(indexed); ... indexed(...)` because the call optrees look different. I wanted to make sure both cases are optimized correctly (for all builtins).

This change is now a separate commit.

tonycoz commented on op.c:
```diff
@@ -9646,6 +9646,35 @@ Perl_newWHILEOP(pTHX_ I32 flags, I32 debuggable, LOOP *loop,
     return o;
 }
 
+#define resolve_lex_sv(t) S_resolve_lex_sv(aTHX_ t)
+static SV *
+S_resolve_lex_sv(pTHX_ PADOFFSET t) {
```
This seems remarkably similar to find_lexical_cv() (which handles some cases this doesn't)

mauke replied:
Ah, thank you. I knew this functionality had to be there somewhere, but I couldn't find anything in perlapi/perlintern and I didn't spot it in the code.

tonycoz commented on pod/perldelta.pod:
```diff
@@ -366,7 +366,29 @@ manager will later use a regex to expand these into links.
 =item *
 
-XXX
+Certain constructs involving a two-variable C<for> loop would crash the perl
+compiler in v5.42.0:
+
```
The SHA the perldelta refers to is going to change once this is merged.

I used to add the SHA in when I did separate PRs/commits after the fix itself was merged, but it doesn't work when the perldelta is included in the PR :(

mauke replied:
Good point. I've changed the commit message.

mauke committed:
perf/opcount.t: ensure imported builtins are optimized

When I saw (in #23429) that

    use builtin qw(indexed);
    sub { for my ($x, $y) (indexed) {} }

crashes, but

    sub { for my ($x, $y) (builtin::indexed) {} }

works fine, I realized that optrees for calls to lexically imported subs
look different from calls to package subs.

This commit make sure both call variants are optimized to direct ops.

tonycoz commented on op.c:
```diff
@@ -9689,10 +9689,17 @@ S_op_is_cv_xsub(pTHX_ OP *o, XSUBADDR_t xsub)
         return false;
 
-    OP *cvop = cLISTOPx(cUNOPo->op_first)->op_last;
+    OP *aop = cUNOPo->op_first;
```
This could probably use a comment like `/* entersub might not be a LISTOP */`

I'll admit I doubted this at first, but there's obvious cases in op.c (anything that calls new_entersubop()) and it's been the case for a long, long time.

mauke replied:
Comment added.

mauke committed:
op.c: factor out repeated entersub code

... for locating the beginning of the argument list and the end (which
is the entered sub itself) in the op tree.

leonerd commented on op.c:
```diff
@@ -9652,6 +9652,33 @@ Perl_newWHILEOP(pTHX_ I32 flags, I32 debuggable, LOOP *loop,
+        aop = cUNOPx(aop)->op_first;
+    }
+    /* move past pushmark */
+    aop = OpSIBLING(aop);
```
I'd be tempted to add an `assert` for this, just to check

```c
assert(OpTYPE(aop) == OP_PUSHMARK);
```

mauke replied:
None of the other parts I cribbed this code from had an `assert` there and I didn't want to make the code "crashier" than before.

leonerd reviewed (approved):
Possibly worth adding an `assert` but otherwise LGTM
